src/osd/scrubber/scrub_reservations.cc

```diff
@@ -125,16 +124,52 @@ ReplicaReservations::~ReplicaReservations()
   log_failure_and_duration(scrbcnt_resrv_aborted);
 }
 
+tl::expected<bool, std::string> ReplicaReservations::is_response_relevant(
+    epoch_t msg_epoch,
+    reservation_nonce_t msg_nonce,
+    pg_shard_t from) const
+{
+  if ((msg_nonce != 0) && (msg_nonce != m_last_request_sent_nonce)) {
+    return tl::unexpected(fmt::format(
+	"stale reservation response from {} (with e:{}) (req-id :{} vs. "
+	"expected {})",
+	from, msg_epoch, msg_nonce, m_last_request_sent_nonce));
+  }
+  return true;
+}
+
+tl::expected<bool, std::string> ReplicaReservations::is_msg_source_correct(
+    epoch_t msg_epoch,
+    pg_shard_t from) const
+{
+  const auto exp_source = get_last_sent();
+  if (!exp_source || from != *exp_source) {
+    return tl::unexpected(fmt::format(
+	"unexpected response from {} (with e:{}) (expected {})", from,
+	msg_epoch, exp_source.value_or(pg_shard_t{})));
+  }
+  return true;
+}
```

Comment:
These helpers are more confusing than helpful.  Including msg_epoch in order to populate the error message suggests to the reader that it's relevant to the predicate.  I'd rather that they be pure predicates (returning bool) with the caller producing the error message.